src/Symfony/Component/Asset/Preload/PreloadManager.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Asset\Preload;
+
+use Symfony\Component\Asset\Exception\InvalidArgumentException;
+
+/**
+ * Manages preload HTTP headers.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class PreloadManager implements PreloadManagerInterface
+{
+    private $resources = array();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addResource($uri, $as = '', $nopush = false)
+    {
+        $this->resources[$uri] = array('as' => $as, 'nopush' => $nopush);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getResources()
+    {
+        return $this->resources;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setResources(array $resources)
+    {
+        foreach ($resources as $key => $options) {
+            if (!is_string($key)) {
+                throw new InvalidArgumentException('The key must be a path to an asset, "%s" given.', $key);
+            }
+
+            if (!isset($options['as']) || !is_string($options['as'])) {
+                throw new InvalidArgumentException('The "as" option is mandatory and must be a string.');
```

Comment:
I have a question about this behavior because I'm not aware of the related specification. If `as` is mandatory, would it make sense to make it optional and default its value to the appropriate value according to the extension of the asset?  For example, if the asset is `foo.js` then use `script` as `as` automatically.